### ISSUE TITLE
User component should respect Accept header.

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#10937]
When using certain kinds of requests, like AJAX, it makes no sense to send a redirect to the login page.
For AJAX requests Yii handles this by throwing a forbidden exception, returning `403`.

Some non-AJAX requests don't accept `text/html` and it doesn't ever make sense to redirect them to the login page.
One such example is an `EventSource` (https://developer.mozilla.org/en-US/docs/Web/API/EventSource).

In conclusion, currently the redirect to the login page is incorrect when `text/html` is not `acceptable` (ie, mentioned explicitly, via `*` or by omission of the `Accept` header).

In such cases it makes more sense to throw a `401 Unauthorized` error than to redirect the user to a page that does not fill the request specifications (and that should actually throw a `406 Not Acceptable` instead of just blindly serving `text/html`.

Side note, for AJAX request, this code makes no sense:
````
throw new ForbiddenHttpException(Yii::t('yii', 'Login Required'));
````
If login is required the code should be `401` not `403`.

**Comment by cebe** [comment]
> If login is required the code should be 401 not 403.

https://github.com/yiisoft/yii2/issues/9846#issuecomment-146891825

**Comment by SamMousa** [comment]
Fair enough for the `401` vs `403` debate. That leaves the `Accept` header open for discussion.

**Comment by SamMousa** [comment]
Can I go ahead and implement this?
--> Meaning AJAX requests and requests with an `Accept` header that does not include: "*", "text/html" or "application/xhtml+xml"
(https://www.w3.org/TR/xhtml-media-types/#application-xhtml-xml)

I will add a configuration setting so developers can adapt this list of acceptable types; since other types could theoretically create a login prompt (maybe flash or some custom media type).

**Comment by cebe** [comment]
I'm not entirely sure how an implementation should look like but I think something like this should be added so you can go ahead writing a proposal.

**Comment by SamMousa** [comment]
@cebe, ok.
Question: currently an AJAX request will not set the return URL, but it will send a redirect. Is that desired behavior? Does it make sense to redirect AJAX requests that generally happen in the background to the login page that the user likely will never see?

**Comment by klimov-paul** [comment]
> Does it make sense to redirect AJAX requests that generally happen in the background to the login page that the user likely will never see?

It make sense. Since you can use `Pjax` or similar solution.

**Comment by SamMousa** [comment]
Ok, won't change that then :)

**Comment by SamMousa** [comment]
Note on my implementation:
Setting `$acceptableRedirectTypes` to `[]` will still allow redirects when the browser omits the `Accept` header or sets it to `*`.

**Comment by SilverFire** [comment]
Fixed in https://github.com/yiisoft/yii2/commit/040e102d97c7af909da829c02825cdd6ff923e0e

**Comment by SilverFire** [comment]
Thank you

**Comment by dmitry-kulikov** [comment]
@SamMousa 
I have question about your changes https://github.com/yiisoft/yii2/commit/040e102d97c7af909da829c02825cdd6ff923e0e#commitcomment-17366730
I'm asking because these changes altered default behavior for ajax request. Lets suppose user is unauthenticated, he does some action which lead to ajax request -
before your changes:
user redirected to Login page;
after your changes:
403 (Forbidden) is shown in browser console, no redirect.
Simplest way to check: install yii2 advanced app template, run built-in PHP server, go to http://localhost:8080/backend/web/index.php?r=site%2Flogin, execute in browser console `$.ajax('http://localhost:8080/backend/web/index.php');` result will differ for yii2 2.0.7 and 2.0.8, with 2.0.7 redirect will happen, no redirect with 2.0.8.
I understand that I can return old behavior by adding `'*/*'` to acceptableRedirectTypes, but is it correct default behavior for Yii? If so then makes sense to add more details in https://github.com/yiisoft/yii2/blob/master/framework/UPGRADE.md#upgrade-from-yii-207, cause I suspect many developers can rely on old behavior and it is not clear that it was changed.

**Comment by SamMousa** [comment]
Will reply in detail in ~5 days.

**Comment by cebe** [comment]
@dmitry-kulikov can you please open a new issue for this so we can tag it properly?

**Comment by dmitry-kulikov** [comment]
https://github.com/yiisoft/yii2/issues/11523

**Comment by DmLapin** [comment]
This change **has broke** our application for some IE clients, that have `*/*` in Accept header. I see fix in #11523 fixes it, but there are high-level problems:

1. Assumption that "redirect to the login page is incorrect when text/html is not acceptable" is wrong. [RFC7231](https://tools.ietf.org/html/rfc7231) doesn't containt such restriction. "Accept" header has no intersections with applicability of redirects. Moreover, you have no way to know whether the client supports redirect or not.

2. Even if you introduce RedirectAcceptable-feature, is \yii\web\User good place for it? yii\web\Request or Response seems to be more appropriate place. Moreover, feature should be disabled by default.

**Comment by SamMousa** [comment]
1. If you specify that you accept just JSON, then redirecting to a page
that has just HTML should not even return content right? Since there is a
separate HTTP code for not acceptable?

**Comment by DmLapin** [comment]
@SamMousa if you specify that you accept only JSON and send request to web-application, that requires authentification through web form, this firstly seems to be an arhitectural mistake. Secondly in this hipotetical situation you should use ContentNegotiator filter or throw `406 Not Acceptable` as HTTP specification says.
But I can't still imagine situation then "non-AJAX requests don't accept text/html". Can you provide example of real application, please? Yes, it can be in API-application, but it must use [another](http://www.yiiframework.com/doc-2.0/guide-rest-authentication.html) authentification scheme. We're talking about something else.
And few words about ajax-requests.
> When using certain kinds of requests, like AJAX, it makes no sense to send a redirect to the login page.

Why do you think so? If browser supports redirect in response on ajax call, it **legal** situation. And it's real usecase for classic web-application -- redirect client to auth form. Developer should be responsible for behaviour of his application. We can use default \yii\filters\AccessControl that triggers `$user->loginRequired()` regardless of the accept header. But if we don't want to do redirect it some cases, we can develop own AccessControl-filter, that will take into account http-headers, ajax-requests, etc.

**Comment by SamMousa** [comment]
The redirection I am talking about is not an HTTP 302 but a custom header
that uses JavaScript to redirect the main window to the URL. There is
nothing standard about that.

If I do an Ajax request, under no circumstance is it expected behavior that
the calling page gets redirected, right?

**Comment by SamMousa** [comment]
Also regarding the earlier argument that `yii\web\User` might not be the correct place is, I believe, incorrect.
Redirection in AJAX is perfectly valid according to the RFC, but redirecting the calling page to an HTML page when JSON is requested is not. Since that is exactly the goal of `loginRequired()` it feels like the correct place to me.

My real world applications often do several AJAX requests in parallel and sometimes people get logged out, rather then redirecting them away from the page I prefer to show them a popup allowing them to login again after which I can resubmit the AJAX request(s).
By returning a proper http response code for the AJAX requests, I can still implement whatever behavior I want using JS (which is exactly the way the old behavior was implemented), but I can also easily override it to do something else.

For API applications this redirection does not make sense at all since there is usually not an interactive login which would require redirecting the API consumer to another page.

When looking at a theoretical level, the whole redirection on login is wrong by the way, @DmLapin. Http 302 means "I found what you requested, but it's over there...", but that is not actually what happens. Of course this pattern is very common, but since you mentioned it is a high-level mistake, I disagree. The common scenario is a high-level mistake (but still supported by Yii).

**Comment by SamMousa** [comment]
> @SamMousa if you specify that you accept only JSON and send request to web-application, that requires authentification through web form, this firstly seems to be an arhitectural mistake. 

It is not, if my application is a normal web application where some data gets loaded via AJAX then it would be very valid for me to only accept JSON in those AJAX requests.

**Comment by DmLapin** [comment]
@SamMousa you interpret my words incorrectly. I agree that AJAX+Redirect _may_ (not _must_) be undesirable. I disagree with the way this change was made. Let me summarise my point:

1. Applicability of redirect doesn't depend on Accept-header.
2. To redirect or not to redirect -- it's custom logic of concrete application. Until this change it was served by components like \yii\filters\AccessControl and \yii\filters\auth\AuthMethod for api-applications.
3. Change was made in breaking compatibility manner. Moreover, it wasn't tested enough: some desktop clients like Internet Explorer send `*/*` in Accept header for usual non-ajax requests. Thanks to our testing department, error was detected before our application release.
4. Change wasn't documented in the upgrade guide.

As for AJAX and "arhitectural mistakes", I mean the following: it's good practice to have separate ajax-connectors with own accept-filters, content types and own reaction on "logged out" state. In this case you will never get redirect to html-page. This also applies to any other cases when text/html is not acceptable.

**Comment by SamMousa** [comment]
> Applicability of redirect doesn't depend on Accept-header.
I agree for normal 302 redirect. But that wasn't the old behavior, the old behavior sent a custom header, right?

> To redirect or not to redirect -- it's custom logic of concrete application. Until this change it was served by components like \yii\filters\AccessControl and \yii\filters\auth\AuthMethod for api-applications.
The `AccessControl` filter uses `yii\web\User` to take action when a login is required, so it makes sense to execute whatever behavior we want from there.
> Change was made in breaking compatibility manner. 
This should have been documented better.
> Moreover, it wasn't tested enough: some desktop clients like Internet Explorer send */* in Accept header for usual non-ajax requests. Thanks to our testing department, error was detected before our application release.
I'm glad it didn't break your production environment! I'm not sure I understand this point though; we explicitly test for that accept header, and in that case behavior is not changed:
````
protected function checkRedirectAcceptable()
    {
        $acceptableTypes = Yii::$app->getRequest()->getAcceptableContentTypes();
        if (empty($acceptableTypes) || count($acceptableTypes) === 1 && array_keys($acceptableTypes)[0] === '*/*') {
            return true;
        }

        foreach ($acceptableTypes as $type => $params) {
            if (in_array($type, $this->acceptableRedirectTypes, true)) {
                return true;
            }
        }

        return false;
    }
````

**Comment by DmLapin** [comment]
> we explicitly test for that accept header, and in that case behavior is not changed:

I talk about [release 2.0.8](https://github.com/yiisoft/yii2/blob/2.0.8/framework/web/User.php#L697-L711), which contains:

```
protected function checkRedirectAcceptable()
    {
        $acceptableTypes = Yii::$app->getRequest()->getAcceptableContentTypes();
        if (empty($acceptableTypes)) {
            return true;
        }

        foreach ($acceptableTypes as $type => $params) {
            if ($type === '*' || in_array($type, $this->acceptableRedirectTypes, true)) {
                return true;
            }
        }

        return false;
}
```

**Comment by SamMousa** [comment]
Ah okay, just checked my local version without realizing that the bug you refer has been fixed by now. So then your comment is mainly about the high-level stuff and documentation, correct?

Regarding the high-level stuff, do you still feel that my logic is invalid? If so, can you tell me which parts exactly?

**Comment by DmLapin** [comment]
> Regarding the high-level stuff, do you still feel that my logic is invalid? If so, can you tell me which parts exactly?

As for me, I see overcomplicated logic in `loginRequired()`, which rely on disputable method `checkRedirectAcceptable()`. If you want to make "polymorfic" method for classical web requests (both ajax and non-ajax), it would better don't parse Accept header and to do something like this:
```
if ($this->loginUrl !== null && (!$checkAjax || !$request->getIsAjax())) {
    // redirect
} else {
   // ForbiddenHttpException
}
```
But this approach is not good too, in my opinion. All discussed cases violates S from SOLID: component User knows too much about external conditions (type of request, request headers, type of login page), while doc-block says: "Note that User only maintains the user authentication status. It does NOT handle how to authenticate a user.". Redirect to login page is a part of "how to authenticate". I think, acces control filter is more appropriate place for any auth/redirect logic, because it works on "request" level. It can be substituted by another filter for non-html requests or it can be polymorfic for classic requests:
```
// just example of hypotetical WebAccesControl
protected function denyAccess($user, $request, $responce)
{
    if ($user->getIsGuest() && !$request->getIsAjax()) {
        $responce->redirect($user->getLoginUrl());
    } else {
        throw new ForbiddenHttpException(Yii::t('yii', 'Login Required'));
    }
}
```
```
// just example of special AjaxAccesControl
protected function denyAccess($user, $request, $responce)
{
    if ($user->getIsGuest()) {
        throw new ForbiddenHttpException(Yii::t('yii', 'Login Required'));
    } else {
        throw new ForbiddenHttpException(Yii::t('yii', 'Insufficient rights'));
    }
}
```

**Comment by SamMousa** [comment]
Hmm, that's an interesting point, but that would mean removing the current `loginRequired()` from `yii\web\User` as a whole. 
While I'm not necessarily against it that is a bit out of scope for these specific changes. 
I chose to implement it in `yii\web\User` since that is where the redirect / checks were being done. 
Note that having this somewhere central makes sense from a DRY perspective though...
